### PR TITLE
Add generateCategoricalChart types: string properties, and events

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

Types!

## Related Issue

https://github.com/recharts/recharts/discussions/3717

## Motivation and Context

I want to refactor this whole thing but I need typescript to work with me

## How Has This Been Tested?

TS, Jest

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes
- [X] All new and existing tests passed.

### CODE REVIEW

File: src/chart/generateCategoricalChart.tsx [hunk 1]
```diff
@@ -1415,6 +1420,7 @@ export const generateCategoricalChart = ({
 
     clearDeferId = () => {
       if (!isNil(this.deferId) && deferClear) {
+        // @ts-expect-error types miss the relationship between deferClear and defer
```
Comment (nikolasrieble) [inline src/chart/generateCategoricalChart.tsx:1423]: What do you suggest we do with these statements? 
Should we create a follow up issue? 

I am a bit sceptical of the value of introducing types if that requires us to also `@ts-expect-error`.

Comment (nikolasrieble) [reply src/chart/generateCategoricalChart.tsx:1423]: Do you see a way how you could still fix these in this PR?

**Review by nikolasrieble — COMMENTED** [review]
Great small PR in the right direction. Curious to hear your thoughts on how to proceed with these types that require `@ts-expect-error`

File: src/chart/generateCategoricalChart.tsx (hunk 1 above)
Comment (PavelVanecek) [reply src/chart/generateCategoricalChart.tsx:1423]: These two are forming a pair and if we want the pair to resolve to the same generic then they need to be in the same generic container - like an object or a Class. I can fix this one.

**Comment by PavelVanecek** [comment]
So about the `@ts-expect-error` in general. In my opinion it is not amazing to have that but it is 100% better than `any` because:

- It shows, explicitly, with a comment, that the types have a problem. Developers are alerted to the presence of the problem, can read what it is, and can apply extra caution.
- All the _other_ currently existing places in code, where our function is used, are now strongly typed. `ts-expect-error` is only applied to one single line.
- If a problem is fixed and the `ts-expect-error` is no longer necessary, typescript will highlight that as an error. On the other hand, `any` will spread silently even when it's no longer needed.
- And most important, all the future uses of code are going to be strongly typed.

So you will see me sometimes replacing `any` with `@ts-expect-error`. I tend to do that more often when the required refactor is too large (out of scope of single PR), or when it touches code we want to get rid of anyway (element cloning).

File: src/util/deferer.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * Will execute callback fn asynchronously.
+ * It will detect the appropriate function to use.
+ *
+ * Named after the famous Swiss tennis player, Roger Deferer.
+ *
+ * @param {Function} callback will be executed asynchronously, with no arguments
+ * @returns {Function} a cancel function.
+ */
+export function deferer(callback: () => void): CancelFunction {
```
Comment (nikolasrieble) [inline src/util/deferer.ts:12]: Nice, thank you.

**Review by nikolasrieble — APPROVED** [review]
🚀

**Pull request merged**